lib/srv/desktop/windows_server.go

```diff
@@ -438,72 +438,86 @@ func (s *WindowsService) Serve(plainLis net.Listener) error {
+	defer proxyConn.Close()
 	log := s.cfg.Log
+	tdpConn := tdp.NewConn(proxyConn)
+
+	// Inline function to enforce that we are centralizing TDP Error sending in this function.
+	sendTdpError := func(message string) {
+		if err := tdpConn.OutputMessage(tdp.Error{Message: message}); err != nil {
+			s.cfg.Log.Errorf("Failed to send TDP error message %v: %v", tdp.Error{Message: message}, err)
+		}
+	}
 
 	// Check connection limits.
-	remoteAddr, _, err := net.SplitHostPort(con.RemoteAddr().String())
+	remoteAddr, _, err := net.SplitHostPort(proxyConn.RemoteAddr().String())
 	if err != nil {
-		log.WithError(err).Errorf("Could not parse client IP from %q", con.RemoteAddr().String())
+		log.WithError(err).Errorf("Could not parse client IP from %q", proxyConn.RemoteAddr().String())
+		sendTdpError("Internal error.")
 		return
 	}
 	log = log.WithField("client-ip", remoteAddr)
 	if err := s.cfg.ConnLimiter.AcquireConnection(remoteAddr); err != nil {
 		log.WithError(err).Warning("Connection limit exceeded, rejecting connection")
+		sendTdpError("Connection limit exceeded.")
 		return
 	}
 	defer s.cfg.ConnLimiter.ReleaseConnection(remoteAddr)
 
 	// Authenticate the client.
-	tlsConn, ok := con.(*tls.Conn)
-	if !ok {
-		log.Errorf("Got %T from TLS listener, expected *tls.Conn", con)
-		return
-	}
-	ctx, err := s.middleware.WrapContextWithUser(s.closeCtx, tlsConn)
+	ctx, err := s.middleware.WrapContextWithUser(s.closeCtx, proxyConn)
 	if err != nil {
 		log.WithError(err).Warning("mTLS authentication failed for incoming connection")
+		sendTdpError("Connection authentication failed.")
 		return
 	}
 	log.Debug("Authenticated Windows desktop connection")
 
 	authContext, err := s.cfg.Authorizer.Authorize(ctx)
 	if err != nil {
 		log.WithError(err).Warning("authorization failed for Windows desktop connection")
+		sendTdpError("Connection authorization failed.")
 		return
 	}
 
 	// Fetch the target desktop info. Name of the desktop is passed via SNI.
-	desktopName := strings.TrimSuffix(tlsConn.ConnectionState().ServerName, SNISuffix)
+	desktopName := strings.TrimSuffix(proxyConn.ConnectionState().ServerName, SNISuffix)
 	log = log.WithField("desktop-name", desktopName)
 
 	desktop, err := s.cfg.AccessPoint.GetWindowsDesktop(ctx, desktopName)
 	if err != nil {
-		log.WithError(err).Warning("Failed to fetch desktop by name")
+		log.WithError(err).Warningf("Failed to fetch desktop by name: %v", desktopName)
+		sendTdpError("Teleport failed to find the requested desktop in its database.")
 		return
 	}
 
 	log = log.WithField("desktop-addr", desktop.GetAddr())
 	log.Debug("Connecting to Windows desktop")
 	defer log.Debug("Windows desktop disconnected")
 
-	if err := s.connectRDP(ctx, log, tlsConn, desktop, authContext); err != nil {
-		log.WithError(err).Error("RDP connection failed")
+	if err := s.connectRDP(ctx, log, proxyConn, desktop, authContext); err != nil {
+		log.Errorf("RDP connection failed: %v", err)
+		sendTdpError("RDP connection failed.")
```

Comment:
If the connection is intentionally closed, does this error show up in the UI?

Reply:
It will show as an error if we intentionally close the connection on the backend, like if I add a

```
time.AfterFunc(5*time.Second, func() { proxyConn.Close() })
```

above here. It won't however if I click "Disconnect" through the UI.

Reply:
I think that's ok, at least for now. I haven't come up with a scenario where we want to close the connection on the backend without either 

1) The close being initiated from the frontend (clicking "Disconnect")
2) Us wanting the frontend to display an error (rdp error or some other error causing the `proxyConn` to close).